torch/_inductor/codegen/triton.py

```diff
@@ -2825,6 +2830,81 @@ def select_tiling(cls, node_schedule, numel, reduction_numel=sympy.Integer(1)):
     def flush(self):
         pass
 
+    def benchmark_fused_nodes(self, nodes):
+        _, (numel, rnumel) = max(nodes, key=lambda x: int(x.is_reduction())).group
+        node_schedule = self.generate_node_schedule(nodes, numel, rnumel)
+        tiled_groups = self.select_tiling(node_schedule, numel, rnumel)
+        reduction_hint_val, mutations, index_dtype = self.get_kernel_args(
+            node_schedule, numel, rnumel
+        )
+
+        kernel = TritonKernel(
+            *tiled_groups,
+            reduction_hint=reduction_hint_val,
+            mutations=mutations,
+            index_dtype=index_dtype,
+        )
+
+        # empty last_usage. May cause more aggressive 'evict_last'. Should be fine.
+        for n in nodes:
+            n.last_usage = set()
+
+        self.codegen_node_schedule_with_kernel(node_schedule, kernel)
+        with config.patch("benchmark_kernel", True), V.set_kernel_handler(kernel):  # type: ignore[attr-defined]
+            src_code = kernel.codegen_kernel()
+
+        src_code = src_code.replace(str(Placeholder.KERNEL_NAME), "triton_")
+        mod = PyCodeCache.load(src_code)
+
+        def cache_file_path():
+            return os.path.splitext(mod.__file__)[0] + ".kernel_perf"  # type: ignore[type-var,operator]
+
+        def load_cache():
+            path = cache_file_path()
+            if os.path.exists(path):
+                with open(path) as fd:
+                    return float(fd.read())
+            return None
+
+        def store_cache():
+            path = cache_file_path()
+            with open(path, "w") as fd:
+                fd.write(str(ms))
+
+        log.debug(
+            "kernel src code for %s written to: %s",
+            {n.get_name() for n in nodes},
+            mod.__file__,
+        )
+        ms = load_cache()
+        if ms is not None:
+            return ms
+
+        args = mod.get_args()
+        call = mod.call
+        wrapped_jit_function = mod.triton_
+
+        # call once to trigger the compilation
+        call(wrapped_jit_function.clone_args(*args))
+
+        launchers = wrapped_jit_function.launchers
+        assert len(launchers) == 1
+        if launchers[0].n_spills > 0:
+            # skip benchmarking the kernel if there are register spills
+            ms = float("inf")
```

Comment:
Sometimes using spills is not actually bad, so perhaps we want to remove this optimisation?

Reply:
Instead, it may be better to simply log that there were register spills and how many

Reply:
hmm, do you have examples that the fused spilled kernel can run faster (when the kernels before fusion does not spill)? I assume even if we have such examples, it would be really rare? So the optimization would still be a good heuristic to prune unnecessary benchmarking work.

Reply:
I don't have any examples, but I think it could happen if the number of spilled registers is not too large. Registers first spill into L1, so if there aren't that many and they fit in L1, spilling registers is not particularly slow. This can get even better if we use cache modifiers like `.cg` when we know we are not going to reuse the data, as to not pollute the L1.

Perhaps we could change this to still benchmark if there are not too many spilled registers?

Reply:
I agree that cache can help a bit to make accessing the spilled register not too slow. Although I don't know how helpful that can be since GPU L1 cache is relatively small.

But more importantly when register start spills, each thread usually already uses 255 registers (the max #of registers a thread can use). That already limit the occupancy to be at most 12.5% .  It's not always but such low occupancy can usually cause bad perf.